lib/scholar/cluster/k_means.ex

```diff
@@ -0,0 +1,325 @@
+defmodule Scholar.Cluster.KMeans do
+  @moduledoc """
+  K-Means algorithm
+  """
+  import Nx.Defn
+  import Scholar.Shared
+
+  @derive {Nx.Container, containers: [:num_iterations, :clusters, :inertia, :labels]}
+  defstruct [:num_iterations, :clusters, :inertia, :labels]
+
+  @doc """
+  Fits a K-Means model for sample inputs `x`
+
+
+  ## Options
+
+    * `:num_clusters` - The number of clusters to form as well as the number of centroids to generate. Defaults to `8`.
+
+    * `:max_iterations` - Maximum number of iterations of the k-means algorithm for a single run. Defaults to `300`.
+
+    * `:num_runs` - Number of time the k-means algorithm will be run with different centroid seeds. The final results will be the best output of num_runs runs in terms of inertia. Defaults to `10`.
+
+    * `:tol` - Relative tolerance with regards to Frobenius norm of the difference in the cluster centers of two consecutive iterations to declare convergence. Defaults to `1e-4`.
+
+    * `:random_state` - Determines random number generation for centroid initialization. Use an int to make the randomness deterministic. The argument `nil` is special and means that the seed is not set. Defaults to `nil`.
+
+    * `:weights` - The weights for each observation in x. If equals to `nil`, all observations are assigned equal weight.
+
+    * `:init` - Method for initialization (Defaults to `:k_means_plus_plus`):
+
+        * `:k_means_plus_plus` -  selects initial cluster centroids using sampling based on an empirical probability distribution of the points’ contribution to the overall inertia. This technique speeds up convergence, and is theoretically proven to be O(log(k))-optimal.
+
+        * `:random` - choose :num_clusters observations (rows) at random from data for the initial centroids.
+
+  ## Returns
+
+    The function returns a struct with the following parameters:
+
+    * `:clusters` - Coordinates of cluster centers.
+
+    * `:num_iterations` - Number of iterations run.
+
+    * `:inertia` - Sum of squared distances of samples to their closest cluster center.
+
+    * `:labels` - Labels of each point.
+  """
+
+  defn fit(x, opts \\ []) do
+    opts =
+      keyword!(
+        opts,
+        num_clusters: 8,
+        max_iterations: 300,
+        num_runs: 10,
+        tol: 1.0e-4,
+        random_state: nil,
+        init: :k_means_plus_plus,
+        weights: nil
+      )
+
+    verify(x, opts)
+    x = x |> Nx.as_type({:f, 32})
```

Comment:
Do we need to forcibly upcast here?

Reply:
Actually, this could downcast if the type is already f64. I would try to avoid this `as_type` here